### PR TITLE
Add Roadmap to Documentation

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1102 suggested by @xudong963 

** All Suggestions Welcome **

 # Rationale for this change
1. Coordinate development work
2. Help Developers new to datafusion to understand where we are going

See also: 
* [mailing list discussion](https://lists.apache.org/thread.html/rdd45ae81adba2d3589752f6c3fd9ead9576e70e2352240e59c9225c1%40%3Cdev.arrow.apache.org%3E)
* https://news.ycombinator.com/item?id=28290616
* [Crowd Sourced Roadmap Jan 2021](https://docs.google.com/document/d/1qspsOM_dknOxJKdGvKbC1aoVoO0M3i6x1CIo58mmN2Y/edit)

# What changes are included in this PR?
Add roadmap to docs published on https://arrow.apache.org/datafusion/

# Are there any user-facing changes?
Docs

### CODE REVIEW

File: docs/source/specification/roadmap.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+this roadmap. However, before submitting a large PR, we strongly
+suggest you read the [before
+starting](https://arrow.apache.org/docs/developers/contributing.html#before-starting)
+recommendations to minimize code review surprises.
+
+# DataFusion
+
+## Vision
+
+DataFusion's goal is to become _the de facto query engine_ of choice
```
[the remark this answers is not in the document]
Comment (Jimexist) [reply docs/source/specification/roadmap.md:41]: It means the query engine of choice by fact

[the remark this answers is not in the document]
Comment (xudong963) [reply docs/source/specification/roadmap.md:41]: Got it.  My poor English needs improving (*/ω＼*)

**Comment by alamb** [comment]
Thanks @Jimexist  and @xudong963 . I'll plan to leave this open for the weekend to give anyone who has more time then to contribute and then merge it early next week

File: docs/source/specification/roadmap.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jimexist) [reply docs/source/specification/roadmap.md]: it's okay :-) @xudong963 it was not english and indeed a latin phrase to begin with.

@alamb i think _de facto_ is a common phrase, one way to make it smoother is to put it in italic (like in many books containing latin expressions).

i'd make probably hammer it home by including more examples inline, even by using footnote (not sure GitHub markdown supports that).

File: docs/source/specification/roadmap.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+4. High performance, erogonomic extensibility points at at every layer
+
+The
+
+## Additional SQL Language Features
+
+- Complete support list on [status](https://github.com/apache/arrow-datafusion/blob/master/README.md#status)
+- Timestamp Arithmetic [#194](https://github.com/apache/arrow-datafusion/issues/194)
+- SQL Parser extension point
+- Support for nested structures (fields, lists, structs)
```
Comment (houqp) [inline docs/source/specification/roadmap.md:56]:
```suggestion
- Support for nested structures (fields, lists, structs) [#119](https://github.com/apache/arrow-datafusion/issues/119)
```

File: docs/source/specification/roadmap.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+## Query Optimizer
+
+- Additional constant folding / partial evaluation [#1070](https://github.com/apache/arrow-datafusion/issues/1070)
+- More sophisticated cost based optimizer for join ordering
+
+## Runtime / Infrastructure
+
+- Better support for reading data from remote filesystems (e.g. S3) without caching it locally
+- Migrate to some sort of arrow2 based implementation (see [milestone](https://github.com/apache/arrow-datafusion/milestone/3) for more details)
+
```
Comment (houqp) [inline docs/source/specification/roadmap.md:69]:
```suggestion
- Add DataFusion to h2oai/db-benchmark [147](https://github.com/apache/arrow-datafusion/issues/147)
- Improve build time [348](https://github.com/apache/arrow-datafusion/issues/348)
```

File: docs/source/specification/roadmap.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+- Run all queries from the TPCH benchmark (see [milestone](https://github.com/apache/arrow-datafusion/milestone/2) for more details)
+
+## Query Optimizer
+
+- Additional constant folding / partial evaluation [#1070](https://github.com/apache/arrow-datafusion/issues/1070)
+- More sophisticated cost based optimizer for join ordering
+
+## Runtime / Infrastructure
+
+- Better support for reading data from remote filesystems (e.g. S3) without caching it locally
```
Comment (houqp) [inline docs/source/specification/roadmap.md:67]:
```suggestion
- Better support for reading data from remote filesystems (e.g. S3) without caching it locally [#907](https://github.com/apache/arrow-datafusion/issues/907) [#1060](https://github.com/apache/arrow-datafusion/issues/1060)
```

File: docs/source/specification/roadmap.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (alamb) [reply docs/source/specification/roadmap.md]: 👍  makes sense. I do think in general it is a good idea to try and minimize 'flowery prose' and idiomatic English is probably the best plan for what have become world wide communities.

File: docs/source/specification/roadmap.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
+\<!---
```
[the remark this answers is not in the document]
Comment (alamb) [reply docs/source/specification/roadmap.md:1]: 🤦  no definitely not intentional. Thank you

**Review by houqp — APPROVED** [review]
thanks @alamb for putting this up.

If I recall correctly, @jorgecarleitao has plans to decouple datafusion into smaller reusable crates too. But if @jorgecarleitao and @andygrove are too busy adding their items right now, we could merge and add those items later as follow up PRs.

@Dandandan do you want to add your tokomak optimizer to the list?

**Comment by houqp** [comment]
also cc @yjshen in case we missed any item needed from your native spark executor work.

File: docs/source/specification/roadmap.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+- SQL Parser extension point [#533](https://github.com/apache/arrow-datafusion/issues/533)
+- Support for nested structures (fields, lists, structs) [#119](https://github.com/apache/arrow-datafusion/issues/119)
+- Remaining Set Operators (`INTERSECT` / `EXCEPT`) [#1082](https://github.com/apache/arrow-datafusion/issues/1082)
+- Run all queries from the TPCH benchmark (see [milestone](https://github.com/apache/arrow-datafusion/milestone/2) for more details)
+
+## Query Optimizer
+
+- Additional constant folding / partial evaluation [#1070](https://github.com/apache/arrow-datafusion/issues/1070)
+- More sophisticated cost based optimizer for join ordering
+
```
Comment (Dandandan) [inline docs/source/specification/roadmap.md:63]: Implement advanced query optimization framework (Tokomak) #440

File: docs/source/specification/roadmap.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+- SQL Parser extension point [#533](https://github.com/apache/arrow-datafusion/issues/533)
+- Support for nested structures (fields, lists, structs) [#119](https://github.com/apache/arrow-datafusion/issues/119)
+- Remaining Set Operators (`INTERSECT` / `EXCEPT`) [#1082](https://github.com/apache/arrow-datafusion/issues/1082)
+- Run all queries from the TPCH benchmark (see [milestone](https://github.com/apache/arrow-datafusion/milestone/2) for more details)
+
+## Query Optimizer
+
+- Additional constant folding / partial evaluation [#1070](https://github.com/apache/arrow-datafusion/issues/1070)
+- More sophisticated cost based optimizer for join ordering
+Implement advanced query optimization framework (Tokomak) #440
```
Comment (xudong963) [inline docs/source/specification/roadmap.md:63]:
```suggestion
- Implement advanced query optimization framework (Tokomak) #440
```

**Comment by yjshen** [comment]
> also cc @yjshen in case we missed any item needed from your native spark executor work.

Thanks, @houqp. I think what I need most is covered by the `Resource Management` section. I'm working on prototyping a memory limit version of `SortExec` currently.

On the Ballista side, I feel Broadcast join is great to add. Besides, we could have a sort-based shuffle writer for memory usage friendly and have a single map output file for each task to avoid creating too many small files when the output partition number is significant.

**Comment by houqp** [comment]
Thank you @alamb for making this happen. I can help deploy the change later today.

Reminder for @jorgecarleitao and @andygrove to add your entries before the upcoming datafusion 6.0.0 release ;)

**Pull request merged**